polkadot/node/network/collator-protocol/src/validator_side_experimental/collation_manager/mod.rs

```diff
@@ -0,0 +1,1361 @@
+					Either::Left(None) => continue,
+					Either::Right(delay) => {
+						let min_delay = maybe_min_delay.get_or_insert(delay);
+						maybe_min_delay = Some(std::cmp::min(*min_delay, delay));
+						continue
+					},
+				};
+
+				// This here may also claim a slot of another leaf if eligible.
+				if self.claim_queue_state.claim_pending_slot(
+					&advertisement.relay_parent,
+					&para_id,
+					advertisement.candidate_hash(),
+				) {
+					let req = self.fetching.launch(&advertisement);
+					requests.push(req);
+					continue
+				} else {
+					gum::warn!(
+						target: LOG_TARGET,
+						?leaf,
+						"Could not claim a slot for the chosen advertisement {:?}",
+						advertisement
+					);
+				}
+			}
+		}
+
+		(requests, maybe_min_delay)
+	}
+
+	pub fn remove_peer(&mut self, peer: &PeerId) {
+		for per_rp in self.per_relay_parent.values_mut() {
+			// No need to reset now the statuses of claims that were pending fetch for these
+			// candidates, or even cancel the futures as the requests will soon conclude with a
+			// network error.
+			per_rp.remove_peer_advertisements(peer);
+		}
+	}
+
+	pub fn remove_peers<'a>(&'a mut self, peers_to_remove: impl Iterator<Item = &'a PeerId>) {
+		// Remove advertisements from these peers.
+		for peer in peers_to_remove {
+			self.remove_peer(peer)
+		}
+	}
+
+	pub async fn completed_fetch<Sender: CollatorProtocolSenderTrait>(
+		&mut self,
+		sender: &mut Sender,
+		res: CollationFetchResponse,
+	) -> CanSecond {
+		let advertisement = res.0;
+		self.fetching.completed(&advertisement);
+
+		let mut reject_info = SecondingRejectionInfo {
+			relay_parent: advertisement.relay_parent,
+			peer_id: advertisement.peer_id,
+			para_id: advertisement.para_id,
+			maybe_output_head_hash: None,
+			maybe_candidate_hash: advertisement.candidate_hash(),
+		};
+
+		let Some(per_rp) = self.per_relay_parent.get_mut(&advertisement.relay_parent) else {
+			gum::debug!(
+				target: LOG_TARGET,
+				hash = ?advertisement.relay_parent,
+				para_id = ?advertisement.para_id,
+				peer_id = ?advertisement.peer_id,
+				"Collation fetch concluded for relay parent out of view"
+			);
+			return CanSecond::No(None, reject_info)
+		};
+
+		per_rp.remove_advertisement(&advertisement);
+
+		let Some(session_info) = self.per_session.get(&per_rp.session_index) else {
+			gum::debug!(
+				target: LOG_TARGET,
+				hash = ?advertisement.relay_parent,
+				para_id = ?advertisement.para_id,
+				peer_id = ?advertisement.peer_id,
+				"Collation fetch concluded for relay parent whose session index is unknown"
+			);
+			return CanSecond::No(None, reject_info)
+		};
+
+		match process_collation_fetch_result(res) {
+			Ok(fetched_collation) => {
+				// It can't be a duplicate, because we check before initiating fetch. For the old
+				// protocol version, we anyway only fetch one per relay parent.
+				per_rp
+					.fetched_collations
+					.insert(fetched_collation.candidate_receipt.hash(), advertisement.peer_id);
```

Comment:
If can_begin_seconding() later returns CanSecond::No, the entry remains. Is there a cleanup path for failed seconding attempts ?